definition_tooling/validator/core.py

```diff
@@ -0,0 +1,114 @@
+import abc
+import json
+from pathlib import Path
+from typing import Union
+
+from definition_tooling.api_errors import DATA_PRODUCT_ERRORS
+from definition_tooling.validator import errors as err
+
+
+class BaseValidator:
+    def __init__(self, path: Union[str, Path]):
+        self.path = Path(path)
+
+    def validate(self):
+        try:
+            spec = json.loads(self.path.read_text(encoding="utf8"))
+        except json.JSONDecodeError:
+            raise err.InvalidJSON(f"Incorrect JSON: {self.path}")
+        except Exception as e:
+            raise err.ValidatorError(f"Failed to validate {self.path}: {e}")
+        self.validate_spec(spec)
+
+    @abc.abstractmethod
+    def validate_spec(self, spec: dict):
+        raise NotImplementedError
+
+
+class DefaultValidator(BaseValidator):
+    def validate_spec(self, spec: dict):
+        if not spec.get("openapi", "").startswith("3"):
+            raise err.UnsupportedVersion
+
+
+def validate_component_schema(spec: dict, components_schema: dict):
+    if not spec["content"].get("application/json"):
+        raise err.WrongContentType
+    ref = spec["content"]["application/json"].get("schema", {}).get("$ref")
+    if not ref:
+        raise err.SchemaMissing(
+            'Request or response model is missing from "schema/$ref" section'
+        )
+    if not ref.startswith("#/components/schemas/"):
+        raise err.SchemaMissing(
+            "Request and response models must be defined in the"
+            '"#/components/schemas/" section'
+        )
+    model_name = ref.split("/")[-1]
+    if not components_schema.get(model_name):
+        raise err.SchemaMissing(f"Component schema is missing for {model_name}")
+
+
+def validate_spec(spec: dict):
+    """
+    Validate that OpenAPI spec looks like a data product definition. For example, that
+    it only has one POST method defined.
+
+    :param spec: OpenAPI spec
+    :raises OpenApiValidationError: When OpenAPI spec is incorrect
+    """
+    if "servers" in spec:
+        raise err.ServersShouldNotBeDefined
+
+    if not spec.get("openapi", "").startswith("3"):
+        raise err.UnsupportedVersion
```

Comment:
The default error message contains the version number and this check also contains the version number, it's highly likely now that one will change one of them and miss the other one, I'd prefer to have the message here as well.